client/lib/wp/sync-handler/index.js

```diff
@@ -0,0 +1,179 @@
+	/**
+	 * Create a SyncHandler instance
+	 *
+	 * @param {Object} [setup] - sync setup
+	 * @param {Function} handler - wpcom handler
+	 *
+	 * @return {Function} sync-handler wrapper
+	 */
+	constructor( setup, handler ) {
+		if ( typeof setup === 'function' ) {
+			handler = setup;
+			setup = {};
+		}
+
+		this.setup = Object.assign( {}, defaults, setup );
+		this.reqHandler = handler;
+		return this.syncHandlerInstance( handler );
+	}
+
+	syncHandlerInstance( handler ) {
+		return ( params, wrappedResponseCallback ) => {
+			// detect and no-sync proxy connection request
+			if ( params.metaAPI && params.metaAPI.accessAllUsersBlogs ) {
+				debug( 'skip - non-sync -proxy-handler request detected' );
+				return this.reqHandler( params, wrappedResponseCallback );
+			}
+
+			const clonedParams = Object.assign( {}, params );
+			const { path } = clonedParams;
+
+			// whitelist barrier
+			if ( ! isWhitelisted( params ) ) {
+				debug( 'not whitelisted: skip %o request', path );
+				return handler( params, wrappedResponseCallback );
+			};
+
+			// response sent flag
+			let responseSent = false;
+
+			// generate an unique resource key
+			const key = this.generateKey( params );
+
+			debug( 'starting to get resource ...' );
+
+			this.retrieveRecord( key, ( err, localRecord ) => {
+				if ( err ) {
+					// @TODO improve error handling here
+					console.error( err );
+				}
+
+				// let's be optimistic
+				if ( localRecord ) {
+					responseSent = true;
+
+					debug( '%o stored(%o). Let\'s be optimistic ...\n', path, localRecord );
+					wrappedResponseCallback( null, localRecord.body );
+				} else {
+					debug( 'No data for %o\n', path );
+				}
+
+				// * background sync process *
+				debug( 'requesting %o to WP.com', path );
+				handler( params, ( resErr, resData ) => {
+					if ( resErr ) {
+						if ( responseSent ) {
+							return;
+						}
+
+						return wrappedResponseCallback( resErr );
+					}
+
+					debug( 'WP.com response %o -> %o', path, resData );
+
+					const isPOSTRequest = clonedParams &&
+						clonedParams.method &&
+						'post' === clonedParams.method.toLowerCase();
+
+					if ( ! isPOSTRequest ) {
+						let storingData = {
+							__sync: {
+								key,
+								synced: new Date().toString(),
+								syncing: false
+							},
+							body: resData,
+							params: clonedParams
+						};
+
+						this.storeRecord( key, storingData, storedErr => {
+							if ( storedErr ) {
+								console.error( storedErr );
+								if ( ! responseSent ) {
+									wrappedResponseCallback( err, resData );
+								}
+							}
+
+							// extra response
+							let _resData = Object.assign(
+								{},
+								resData,
+								{ __syncResponse: true }
+							);
+
+							debug( 'background response sent' );
+							wrappedResponseCallback( err, _resData );
+						} );
+					} else {
+						debug( 'skip - non sync request: [%o] %o - %o\n', path, params, resData );
+					}
+
+					if ( ! responseSent ) {
```

Comment:
This is where we have the canonical record back from the server, so this is where I would expect to run the double-callback with the canonical data. So why would we have the check for responseSent here? This is where I expect to run the callback again with the correct data.